indigo_app/forms/works.py

```diff
@@ -480,6 +508,108 @@ def clean(self):
+
+    def __init__(self, work, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.work = work
+
+    @cached_property
+    def commencing_work_obj(self):
+        return Work.objects.filter(pk=self['commencing_work'].value()).first()
+
+    @cached_property
+    def commenced_work_obj(self):
+        return Work.objects.filter(pk=self['commenced_work'].value()).first()
+
+    def save(self, user, *args, **kwargs):
+        commencement = Commencement.objects.filter(pk=self.cleaned_data['id']).first()
+
+        if self.cleaned_data.get('DELETE'):
+            if commencement:
+                commencement.delete()
+            return
+        else:
+            if commencement:
+                commencement.commencing_work = self.cleaned_data['commencing_work']
+                commencement.commenced_work = self.cleaned_data['commenced_work']
+                commencement.date = self.cleaned_data["date"]
+                commencement.note = self.cleaned_data["note"]
+                commencement.updated_by_user = user
+                commencement.save()
+            else:
+                Commencement.objects.create(
+                    commencing_work=self.cleaned_data['commencing_work'],
+                    commenced_work=self.cleaned_data['commenced_work'],
+                    date=self.cleaned_data["date"],
+                    note=self.cleaned_data["note"],
+                    created_by_user=user,
+                    updated_by_user=user,
+                )
+
+CommencementsFormset = formset_factory(
+    CommencementsPartialForm,
+    extra=0,
+    can_delete=True,
+)
+
+class CommencementsMadeBaseFormset(CommencementsFormset):
+
+
+    def clean(self):
+        super().clean()
+        if any(self.errors):
+            return
+        # check if commencing work and date are unique together
+        seen = set()
+        for form in self.forms:
+            if form.cleaned_data.get('DELETE'):
+                continue
+            amending_work = form.cleaned_data.get('commencing_work')
+            amended_work = form.cleaned_data.get('commenced_work')
+            date = form.cleaned_data.get('date')
+            if (amending_work, amended_work, date) in seen:
+                raise ValidationError("Commenced work and date must be unique together.")
+            seen.add((amending_work, amended_work, date))
+
+    def save(self, user, work, *args, **kwargs):
+        if self.is_valid() and self.has_changed():
+            for form in self.forms:
+                form.save(user)
+
+            commenced_counts = Commencement.objects.filter(commencing_work=work).values("commenced_work").annotate(num_commencements=Count("id")).order_by()
+            multiple_commencements = commenced_counts.filter(num_commencements__gt=1).values_list("commenced_work", flat=True)
+            single_commencements = commenced_counts.filter(num_commencements=1).values_list("commenced_work", flat=True)
```

Comment:
so here we should also be taking into account any other existing commencements on all the commenced works, as you rightly pointed out in your video.
- if any of those other commencements is already `main`, don't set any of ours to `main`. if none are, we can still use the earliest (or only) one that we're adding
- if any have `all_provisions=True`, we should set that to False and save that commencement, and make ours False as well

you can probably treat all 'multiple' and 'single' commencements the same, as long as 
- the single and multiple lists are based on all commencements on the commenced work (Commencement.filter(commenced_work=commenced_work))
- the multiples logic is tweaked to check for existing `main` and `all_provisions`

Reply:
@goose-life I've updated the PR with our changes and its working ok. See the demo https://www.loom.com/share/0a35e3c039d948deb5eb802708e8c4fd?sid=73bc0379-b580-4c2e-b4c3-2843677f8f71